# Profitable free apps from GooglePlay and AppStore.
In this project we will analyze using python programming language skills, a set of free apps from the AppStore and GooglePlay. Since free apps depend solely on the number of visitors to generate revenue, is important to understand how to optimize them to atract more traffic and have a better income. 

In [96]:
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
dataset1=list(read_file)
ios_header=dataset1[0]
ios=dataset1[1:]

In [97]:
opened_file = open('googleplaystore.csv')
from csv import reader
read_file = reader(opened_file)
dataset2 = list(read_file)
android_header=dataset2[0]
android=dataset2[1:]

*At the following cell, we will build a function to explore a little bit the content of the datasets we are currently analyzing*

In [98]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

#This function will help us explore the data sets which we are going to use for this project 

*We will now use this function to visualize the two data sets we have.*

In [99]:
explore_data(dataset1,0,4,rows_and_columns = False)

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']




In [100]:
explore_data(dataset2,0,4,rows_and_columns = False)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']




*Firstly we see that `both data sets have a header row`, the first data set has a header row that includes, '','id','track_name',..... and the second data set has a header row that includes, 'App','Category','Rating',....*

*We are going to find the number of rows and the number of columns that each of our data sets have, `taking into account that the header row is not considered to be a row`, so then we apply the following code to get the respective sizes.* 


In [101]:
n_of_rows_dataset1=len(ios)
n_of_rows_dataset2=len(android)
n_of_columns_dataset1=len(ios[0])
n_of_columns_dataset2=len(android[0])
print('The first dataset, whose entries come from the AppleStore, has',n_of_rows_dataset1,'rows,and',n_of_columns_dataset1,'columns')
print('The first dataset, whose entries come from the GooglePlay store, has',n_of_rows_dataset2,'rows,and',n_of_columns_dataset2,'columns')

The first dataset, whose entries come from the AppleStore, has 7197 rows,and 17 columns
The first dataset, whose entries come from the GooglePlay store, has 10841 rows,and 13 columns


Lets see that information summarized in a table 

| Dataset  |rows|columns|
|--------- |----|-------| 
|ios|7197|17|
|GooglePlay|10841|13|


*For the first data set, whose information come from the AppleStore as you've probably already seen the names of the columns are the following*

In [102]:
print(ios_header)


['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


*And for the second data set, whose information come from the GooglePlay store the names of the columns are the following*

In [103]:
print(android_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


*At a quick glance at our GooglePlay dataset, the columns that might be useful for the purpose of our analysis are 'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'.
Now when taking at look at the second dataset whose information come from the AppsStore, we see that the most relevant information for our analysis comes from the following columns, 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre', not all column names are self-explanatory, but for details about the columns of the GooglePlay dataset you can refer to [this link](https://www.kaggle.com/lava18/google-play-store-apps/home) and for more information on the columns of the AppStore dataset you can refer to [this link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home#appleStore_description.csv).*

## Deleting wrong data
At the [link](https://www.kaggle.com/lava18/google-play-store-apps) where the data was originally posted, there is a [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), and I want to focus our attention at [this discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015). For that, on the following code, we will
1. Print the wrong row.
2. Print the header row.
3. Print a non-defective list.

In [104]:
print(android[10472])   # incorrect row
print('\n')
print(android_header)  # header
print('\n')
print(android[0])  # Non-defective rows

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


The first thing you see, once the 3 rows are printed is that the *category* row, is not present, therefore, we have a "1.9" at the place where we should have the category of the app. Therefore we will delete the row. 

In [105]:
print(len(android))
del android[10472]
print(len(android))

10841
10840


## Removing duplicates from the list
When we analyze our data we do not want to use apps that are repeated in the list. Lets see a code that shows that the "Instagram" is repeated **four times** in the list. 

In [106]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)
        print('\n')

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']




Notice that Instagram is not the only App that occurs more than once in our list, if we execute the same code for 'Facebook', we see that 'Facebook' is repeated twice. 

In [107]:
for app in android:
    name = app[0]
    if name == 'Facebook':
        print(app)
        print('\n')


['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']


['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']




Therefore we now want to write a code that will find the total number of repeated apps in our list, therefore we will write a code that does the following. 

Using a for loop our code:

1. Will add the name of the repeated apps to a list called `duplicate_apps` 
2. Will add the name of the non repeated apps to a list called `unique_apps`

In [108]:
duplicate_apps=[]
unique_apps=[]

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
print('The total number of repeated apps is', len(duplicate_apps))
print('\n')
print('Here are some examples of repeated apps', duplicate_apps[:5])

The total number of repeated apps is 1181


Here are some examples of repeated apps ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


Since we do not want to have the same app twice in our analysis we could simply remove the apps that are repeated randomly, but we will take a closer look at the repeated app "Instagram" to see that randomness is not convinient for the deletion of repeated apps. 

In [109]:
for app in android:
    name = app[0]
    if name =='Instagram':
        print(app)
        print('\n')
        

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']




Notice, that each of the lists has a different number of reviews therefore **we want the maximum number of reviews**, since that will be reflect more clearly the customer's opion. 

## Deleting repeated apps

As it was specified above, we want to have the highest number of apps reviews therefore we will make a dictionary that has the following features. 
1. Each dictionary key will be a unique app name. 
2. The value of each key will be the maximum from all reviews available from the original list.


In [110]:
reviews_max={}
for app in android:
    name=app[0]
    n_reviews=float(app[3])
    if name in reviews_max  and reviews_max[name]< n_reviews:
        reviews_max[name]=n_reviews
    elif name not in reviews_max:
        reviews_max[name]=n_reviews        

*We want to make sure that our dictionary is correct, to check that out notice that*

In [111]:
print("The expected length of the dictionary is:", len(android)-len(duplicate_apps))
print("The actual length of the dictionary is:", len(reviews_max))

The expected length of the dictionary is: 9659
The actual length of the dictionary is: 9659


**Therefore the dictionary is correct**

# Using the created dictionary to delete repeated apps. 

Remember that our criteria, to delete repeated apps was to select those apps with the highest number of reviews, but **some of the repeated apps have the same number of reviews**, therefore we need to be careful to avoid repeating apps again. Here is an example of an app, whose name is "Box" and has exactly the same number of reviews every time is repeated.

In [112]:
for app in android: 
    name=app[0]
    if name=="Box":
        print(app)
        print("\n")

['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']




Taking into account what I just discussed above regarding the repeating apps we will write a code that does the following: 
* Initially, two lists named **android_clean**  and  **android_added** will be created. 
* After that, we will loop through the **android**  list.
  - At each iteration the name and the number of reviews will be assigned each to a variable 
  - If the aready assigned variable matches the number of reviews from our dictionary **AND**  the name does not belong to the already added list. The entire list will be appended to the **android_clean** list, and the name of the app will be appended to the **already_added** list. 
  - Notice that the condition "the name does not belong to the already added list" guarantee that we will not repeat an app if it has the same number of reviews each time it repeats. 



In [113]:
android_clean=[]
already_added=[]
for app in android:
    name=app[0]
    n_reviews=float(app[3])
    if ( n_reviews == reviews_max[name] ) and ( name not in already_added ):
        android_clean.append(app)
        already_added.append(name)

And now we will make sure that our clearned data set has the desired size, which is `9,659 rows` and we will explore the data a little bit as well. 

In [114]:
explore_data(android_clean,0,3,rows_and_columns = False)
print('\n')
print("The expected number of rows for our dataset is:", 9659)
print('\n')
print('The actual number of rows of our dataset is:', len(android_clean))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']




The expected number of rows for our dataset is: 9659


The actual number of rows of our dataset is: 9659


**Now we are sure that the our dataset has the right number of apps so we can proceed.**

# Removing apps directed towards a non-English audience 

We want to analyze apps that are directed towards an English speaking audience, in order to filter apps that are only made for English speakers we will have to understand a little bit what Python understands when our input is an string made out of characters. 

- When we type a letter or symbol, using the function `ord` we can see that there is a number associated to it, as we will see now in the following example where we will type some English and non-English characters and find out the number associated to them.

In [115]:
print('The number associated to the letter A is: ', ord("A"))
print("\n")
print('The number associated to the letter a is: ', ord("a"))
print("\n")
print('The number associated to the letter B is:', ord("B"))
print("\n")
print('The number associated to the letter b is:', ord("b"))
print("\n")
print('The number associated to the chinese character 验 is:', ord("验"))
print("\n")
print('The number associated to the ) symbol is:',ord(")"))

The number associated to the letter A is:  65


The number associated to the letter a is:  97


The number associated to the letter B is: 66


The number associated to the letter b is: 98


The number associated to the chinese character 验 is: 39564


The number associated to the ) symbol is: 41


According to the ASCII (American Standard Code for Information Interchange) the English language is written with character ranging from 0 to 127. So we will write code to find out what are the non-English apps in both of our lists by means of a dictionary that will have the name of the app as the key and its respective index in the list asthe value of the key. 

Notice as well that strings are "indexable" and "iterable" as you can see in the following code. 


In [116]:
string = '我会说中文但是我的中文还没有流利'
for character in string:
    print(character)    

我
会
说
中
文
但
是
我
的
中
文
还
没
有
流
利


To filter the words the apps that do not have English characters we will write code in wich:
- An empty dictionary called `non_english_ios` will be created
- A variable called `index = 0` will be created
  * A loop over all the ios apps will be executed
  * For each app its name will be assigned to a variabled called `name` 
  * A loop over the characters of the variable `name` will be executed 
    - If any of the numbers assigned to the characters in the variable name is higher than 127, the app will be added to the dictionary as a key whose value is the index 
- index will be updated by summing one.    

A similar code to find non-English apps in android will be executed 

In [117]:
non_english_ios={}
index=0
for app in ios:
    name=app[2]
    for character in name:
        if ord(character)>127:
            non_english_ios[name]=index

    index+=1   
    


In [118]:
non_english_android={}
index=0
for app in android:
    name = app[0]
    for character in name:
        if ord(character)>127:
            non_english_android[name]=index
    index+=1
    


In [119]:
non_english_ios



{'Google – Search made just for mobile': 12,
 'Lifesum – Inspiring healthy lifestyle app': 25,
 'iHeartRadio – Free Music & Radio Stations': 35,
 'Line Rider iRide™': 36,
 'Lose It! – Weight Loss Program and Calorie Counter': 59,
 'Chase Mobile℠': 67,
 '大辞林': 68,
 '新浪新闻-阅读最新时事热门头条资讯视频': 70,
 'Citi Mobile®': 80,
 'Kindle – Read eBooks, Magazines & Textbooks': 82,
 '同花顺-炒股、股票': 84,
 'Match™ - #1 Dating App.': 94,
 'Brain Wave ™ - 32 Advanced Binaural Brainwave Entrainment Programs with iTunes Music and Relaxing Ambience': 105,
 'LEDit – The LED Banner App': 110,
 "20 Minutes.fr - l'actualité en continu": 117,
 'World Cup Table Tennis™': 124,
 'iStudiez Pro – Homework, Schedule, Grades': 126,
 'MindNode – Delightful Mind Mapping': 134,
 'Guess My Age \ue020 Math Magic': 154,
 'FACEinHOLE®': 167,
 'Pixel Starships™ : 8Bit MMORPG': 170,
 'PGA Championship 2016 – Baltusrol Golf Club': 180,
 'Geocaching®': 200,
 'Sudoku⁺': 204,
 '央视影音-海量央视内容高清直播': 210,
 'Walgreens – Pharmacy, Photo, Coupons a

In [120]:
non_english_android

{'U Launcher Lite – FREE Live Cool Themes, Hide Apps': 2,
 'CarMax – Cars for Sale: Search Used Car Inventory': 85,
 'AutoScout24 Switzerland – Find your new car': 88,
 'Zona Azul Digital Fácil SP CET - OFFICIAL São Paulo': 89,
 'Wattpad 📖 Free Books': 4715,
 'ReadEra – free ebook reader': 161,
 'Docs To Go™ Free Office Suite': 292,
 'USPS MOBILE®': 195,
 'Invoice 2go — Professional Invoices and Estimates': 283,
 'Röhrich Werner Soundboard': 300,
 'Manga Net – Best Online Manga Reader': 303,
 'Truyện Vui Tý Quậy': 309,
 'Comic Es - Shojo manga / love comics free of charge ♪ ♪': 310,
 '漫咖 Comics - Manga,Novel and Stories': 313,
 'Tapas – Comics, Novels, and Stories': 318,
 '【Ranobbe complete free】 Novelba - Free app that you can read and write novels': 321,
 'Messenger – Text and Video Chat for Free': 4104,
 'Yahoo Mail – Stay Organized': 4800,
 'Call Free – Free Call': 349,
 'Xperia Link™': 361,
 'Dolphin Browser - Fast, Private & Adblock🐬': 418,
 'DU Browser—Browse fast & fun': 8901,


In fact we see that there are many non_English apps, so we will create a function that returns false when the input is an string that has a character with al least one non-English character. 

In [121]:
def is_english(x):
    for character in x:
        if ord(character)>127:
            return False
    return True
    
print(is_english('Instagram'))      
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))


True
False


The goal of this section is to remove the non_Eglish apps from our list, the problem is that the codes we have written so far do not take into account that there could be English apps that have symbols that fall out of ASCII range. In the following example we see an English app that is filtered for our function as non-English. 

In [122]:
print(is_english('Bacterial Vaginosis 🇺🇸'))

False


Therefore we will modify our function a little bit and classify apps with more than three characters falling outside of the ASCII range as non_English apps and that will prevent us from loosing big amounts of information. 

In [123]:
def is_english(x):
    vector=[]
    for character in x:
        if ord(character)>127:
            vector.append(character)
    if len(vector)>3:
        return False 
    else:
        return True        
    
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Our function is not perfect but it will help us better filter English apps from non-English apps. We will now use the function we just built to create two new data sets that contain English apps only. 

In [124]:
android_english=[]
ios_english=[]
for app in android_clean:
    if is_english(app[0]):
        android_english.append(app)

for app in ios:
    if is_english(app[2]):
        ios_english.append(app)

explore_data(android_english,0,3,True)
print("\n")
explore_data(ios_english,0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188

We see that we are left with 9614 apps from Google Play and 6183 from the App Store for our analysis.

# Removing non-free apps from the list. 


As we mentioned at the introduction of the project, we will focus only on free apps, so we will write a code to remove all paid apps from our list. If we explore the *android_english* data set we notice that free apps have the string "free" in the type section and if we explore the *ios_english* data set we notice that free apps have *0* in the price entry. 

In [125]:
print(android_header)
print('\n')
print(android_english[100])
print('\n')
print(ios_header)
print('\n')
print(ios_english[1])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Hairstyles step by step', 'BEAUTY', '4.6', '4369', '14M', '100,000+', 'Free', '0', 'Everyone', 'Beauty', 'July 25, 2018', '1.9', '4.0.3 and up']


['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


We will now use that information to filter in a list all free apps. 

In [126]:
android_final=[]
ios_final=[]
for app in android_english:
    if app[6]=='Free':
        android_final.append(app)

for app in ios_english:
    if float(app[5]) == 0:
        ios_final.append(app)

print(explore_data(android_final,0,3,True))
print('\n')
print(explore_data(ios_final,0,0,True))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 8863
Number of columns: 13
None


Number of rows: 3222
Number of columns: 17
None


We see that we end up having 8863 Android apps and 3222 iOS apps after we remove the paid apps. 

# Frequency tables of genres 

At our company we follow a validation strategy that is composed of the following steps when a new application is released. 

1. A first version of the app is released for Android. 
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after 6 months, we develop and iOS version. 

Taking into account that our goal is to develop our app for both markets we want to understand what are succesful profiles of apps in both markets, that is the reason because a frequency table for the `prime_genre` from the iOS data set, and a frequency table of the `Genre`  and `Category` from the android data set will be built. 

The ideal object to be created should be a frequency table and for our analysis we wil have to sort the results using the built-in `sorted()` function but the function does not work very well with a dictionary because is going to try to sort the keys of dictionary instead of sorting the values of the dictionary and the function would work better with tuples as we can see on the following examples. 



In [127]:
a_list=[50,20,100]
print(sorted(a_list))
print('\n')
print(sorted(a_list, reverse = True))

[20, 50, 100]


[100, 50, 20]


As it is clear from the example above the `sorted()` function works very well with a list, now, lets see that it does not work very well with a dictionary. 

In [128]:
a_dictionary={'Genre_1':50,'Genre_2':20,'Genre_3':100}
print(sorted(a_dictionary))
print('\n')
print(sorted(a_dictionary,reverse = True))

['Genre_1', 'Genre_2', 'Genre_3']


['Genre_3', 'Genre_2', 'Genre_1']


The result we were expected to have was `['Genre_2','Genre_1','Genre_3']`  for the application of the `sorted()` function, and similarly the result we got for `reverse =  True` was not correct either. 

Now, notice that the `sorted()` function does work very well with tuples, as you can see on the following example. 

In [129]:
a_tuple = [(50,'Genre_1'),(20,'Genre_2'),(100,'Genre_3')]
print(sorted(a_tuple))
print('\n')
print(sorted(a_tuple, reverse = True))

[(20, 'Genre_2'), (50, 'Genre_1'), (100, 'Genre_3')]


[(100, 'Genre_3'), (50, 'Genre_1'), (20, 'Genre_2')]


Therefore, we are going to write code that has the following features.  
1. Initially, given a data set which is expected to be a list of lists, and an index, a frequency table will be created. 
2. Once the frequence table has been created a function that converts the dictionary into tuples will use the `sorted()` function to print the final result. 

In [130]:
def freq_table(dataset,index):
    dictionary={}
    for entry in dataset:
        if entry[index] not in dictionary:
            dictionary[entry[index]]=1
        else:
            dictionary[entry[index]]+=1
    for key in dictionary:
        dictionary[key]=(dictionary[key]/len(dataset))*100
    return dictionary    



def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

We will now use the two functions just created to display the frequency tables of `prime_genre`, `Genre` and `Category`. 

In [131]:
print("Here the frequency tables of the 'prime_genre' column from the ios_final list of lists will be displayed")
print('\n')
display_table(ios_final,12)

Here the frequency tables of the 'prime_genre' column from the ios_final list of lists will be displayed


Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


The first fact which is very obvious is that the 'Games' category is predominant among the English free apps of the ios market by a wide margin because the following category (which is Entertainment) is just about 7% of the market, while the first most common type of free app (Games) is 58% of the English free apps, there is a difference of about 50% between just the first and the second place. Then we see that Education apps (4th in our ranking) are only about 3%  and Social Networking apps (5th in our ranking) are only about 3% as well. 

Nevertheless; it is widely known that Social Networking apps have a very high number of users while gaming apps tend to have less, which lead us to the conclussion that the demand is not the same as the offer and even though gaming apps are about 58% percent of the market those are not the apps that necesarily drive the most traffic. 

In [132]:
print('Here the frequency table for the "category" column from the adroid_final list of lists will be displayed')
print('\n')
display_table(android_final,1)

Here the frequency table for the "category" column from the adroid_final list of lists will be displayed


FAMILY : 18.898792733837304
GAME : 9.725826469592688
TOOLS : 8.462146000225657
BUSINESS : 4.592124562789123
LIFESTYLE : 3.9038700214374367
PRODUCTIVITY : 3.8925871601038025
FINANCE : 3.7007785174320205
MEDICAL : 3.5315355974275078
SPORTS : 3.396141261423897
PERSONALIZATION : 3.317161232088458
COMMUNICATION : 3.2381812027530184
HEALTH_AND_FITNESS : 3.0802211440821394
PHOTOGRAPHY : 2.944826808078529
NEWS_AND_MAGAZINES : 2.798149610741284
SOCIAL : 2.6627552747376737
TRAVEL_AND_LOCAL : 2.335552296062281
SHOPPING : 2.245289405393208
BOOKS_AND_REFERENCE : 2.1437436533904997
DATING : 1.8616721200496444
VIDEO_PLAYERS : 1.7939749520478394
MAPS_AND_NAVIGATION : 1.399074805370642
FOOD_AND_DRINK : 1.241114746699763
EDUCATION : 1.1621347173643235
ENTERTAINMENT : 0.9590432133589079
LIBRARIES_AND_DEMO : 0.9364774906916393
AUTO_AND_VEHICLES : 0.9251946293580051
HOUSE_AND_HOME : 0.8236488773552973

At a first glance one could think that the ios market and the Googleplay market vary significantly since the apps on the GooglePlay market seem to be more related to everyday tasks, since the most common category we find is "Family" and we find "Tools","Business" and "Lifestyle"  heading the raking, besides, the gap between top places seem to be smaller therefore at first glimps it should be concluded that the iOS and GooglePlay market behave differently. 

Nevertheless; that conclussion is not completely true because of the miss use of terminology. The 'Family' category, is just a different way to categorize games for kids thus summing the percentages of "Games" plus "Family" we obtain that there are actually 26% percent of game-related apps, which is about half from the percentage of the ios market, and the gap is certainly smaller, but we see as well that theres a big predonimation of game apps in the GooglePlay market as well. 

We certainly conclude that even though the GooglePlay market is predenominated by game-realted apps as well there is a much higher percentage of apps for the everyday life, including 'Tools', 'Education','Business', etc. which is confrimed by the 'Genre' frequence table which we print below. 

In [133]:
print('Finally we will display the frequency table for the "Genre" column of the android_final data set.')
print('\n')
display_table(android_final,9)

Finally we will display the frequency table for the "Genre" column of the android_final data set.


Tools : 8.450863138892023
Entertainment : 6.070179397495204
Education : 5.348076272142616
Business : 4.592124562789123
Productivity : 3.8925871601038025
Lifestyle : 3.8925871601038025
Finance : 3.7007785174320205
Medical : 3.5315355974275078
Sports : 3.463838429425702
Personalization : 3.317161232088458
Communication : 3.2381812027530184
Action : 3.102786866749408
Health & Fitness : 3.0802211440821394
Photography : 2.944826808078529
News & Magazines : 2.798149610741284
Social : 2.6627552747376737
Travel & Local : 2.324269434728647
Shopping : 2.245289405393208
Books & Reference : 2.1437436533904997
Simulation : 2.042197901387792
Dating : 1.8616721200496444
Arcade : 1.8503892587160102
Video Players & Editors : 1.771409229380571
Casual : 1.7601263680469368
Maps & Navigation : 1.399074805370642
Food & Drink : 1.241114746699763
Puzzle : 1.128286133363421
Racing : 0.9928917973598104
Role Playi

Now that we have printed, the "Genre" frequency table we see that the main difference between 'Genre' and 'Cateogory' is that 'Genre' is a little bit more precise in terms of classification, and we are looking for a more general overview of the landscape, therefore we will reject the frequency-table for now and then continue analyzing mainly the 'category' frequency table. 

Other conclusion which we could draw from our previous analysis is that the fact that a category is predominant in any of the markets does not necesarily imply that is drives a high amount of traffic. Therefore we will now analyze that 

# Most popular apps by genre on the AppStore

One way to know what are the most popular apps in the AppstStore is by looking at the number of installs, but that information is missing from the the AppleSore data set therefore we will analyze the information, using the the number of reviews per genre.  Which we do down below. 

In [145]:
genres_ios = freq_table(ios_final, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[6])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Productivity : 21028.410714285714
Weather : 52279.892857142855
Shopping : 26919.690476190477
Reference : 74942.11111111111
Finance : 31467.944444444445
Music : 57326.530303030304
Utilities : 18684.456790123455
Travel : 28243.8
Social Networking : 71548.34905660378
Sports : 23008.898550724636
Health & Fitness : 23298.015384615384
Games : 22788.6696905016
Food & Drink : 33333.92307692308
News : 21248.023255813954
Book : 39758.5
Photo & Video : 28441.54375
Entertainment : 14029.830708661417
Business : 7491.117647058823
Lifestyle : 16485.764705882353
Education : 7003.983050847458
Navigation : 86090.33333333333
Medical : 612.0
Catalogs : 4004.0


The highest amount of reviews comes from the "Navigation" genre, but that is highly influenced by GoogleMaps and Waze. 

In [154]:
categories_android=freq_table(android_final,1)
for genre in categories_android:
    total=0
    len_category=0
    for app in android_final:
        category_app=app[1]
        if category_app == genre:
            n_installs=app[5]
            n_installs=n_installs.replace('+','')
            n_installs=n_installs.replace(',','')
            total+=float(n_installs)
            len_category+=1
    avg_n_installs=total/len_category
    print(genre,':',avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3697848.1731343283
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

## Conclusions
In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.
We concluded that practical apps would be profitable in both markets. The markets do vary but when we look at the overall landscape, we can see that practical apps are the best nitch for our purposes, besides, we will avoid developing game apps, since both markets are crowded of game apps. 